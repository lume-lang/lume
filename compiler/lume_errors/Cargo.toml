[package]
name = "lume_errors"
version.workspace = true
edition = "2024"
rust-version.workspace = true

[dependencies]
lume_span = { path = "../lume_span" }

error_snippet = { workspace = true }
error_snippet_derive = { workspace = true }
