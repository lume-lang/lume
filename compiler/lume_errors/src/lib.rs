#![allow(clippy::arc_with_non_send_sync)]
#![feature(negative_impls)]

use error_snippet::{DiagnosticHandler, Handler, Renderer, SimpleDiagnostic};
use std::sync::{
    Arc, Mutex, MutexGuard,
    atomic::{AtomicBool, Ordering},
};

/// Derived from the Rust compiler.
///
/// https://github.com/rust-lang/rust/blob/c79bbfab78dcb0a72aa3b2bc35c00334b58bfe2e/compiler/rustc_span/src/fatal_error.rs
#[derive(Debug, Clone, Copy)]
pub struct FatalError;

pub struct FatalErrorMarker;

// Don't implement Send on FatalError. This makes it impossible to `panic_any!(FatalError)`.
// We don't want to invoke the panic handler and print a backtrace for fatal errors.
impl !Send for FatalError {}

impl FatalError {
    pub fn raise() -> ! {
        std::panic::resume_unwind(Box::new(FatalErrorMarker))
    }
}

impl std::fmt::Display for FatalError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "fatal error occured, aborting...")
    }
}

impl std::error::Error for FatalError {}

/// Defines the different options for outputting diagnostics,
/// once they've been drained from the the diagnostic context ([`DiagCtx`]).
pub enum DiagOutputFormat {
    /// The output should be graphical, i.e. should be forwarded
    /// to the user, so errors and warnings are easy to understand and read.
    Graphical,

    /// No output will be rendered and all diagnosticsw will be sent to the void.
    Stubbed,
}

/// A struct for handling diagnostics.
///
/// The struct is only defined on [`DiagCtx`]-instances, and is only meant to
/// be shared with handles created by them (via the [`DiagCtx::handle()`] and [`DiagCtx::with()`] methods).
struct DiagCtxInner {
    /// The inner handler for diagnostics, which holds all the
    /// reporting diagnostics.
    handler: DiagnosticHandler,

    /// Defines whether the context has reported any errors upon draining
    /// and should exit at the soonest possible convenience.
    tainted: AtomicBool,

    /// Defines whether to halt the application, if an error is encountered
    /// during a drain.
    exit_on_error: bool,
}

impl DiagCtxInner {
    /// Creates a new [`DiagCtxInner`] instance using the given output format.
    fn new(fmt: DiagOutputFormat) -> Self {
        let renderer: Box<dyn Renderer + Send + Sync> = match fmt {
            DiagOutputFormat::Graphical => Box::new(error_snippet::GraphicalRenderer::new()),
            DiagOutputFormat::Stubbed => Box::new(StubRenderer {}),
        };

        let mut handler = error_snippet::DiagnosticHandler::with_renderer(renderer);

        // If we drain an error to the handler, propagate the error count upwards, so
        // we can handle it at a higher level, instead of exiting the application.
        handler.exit_on_error();

        DiagCtxInner {
            handler,
            tainted: AtomicBool::new(false),
            exit_on_error: false,
        }
    }

    /// Drains the currently reported errors in the context to the output buffer.
    ///
    /// For more information, read the documentation on [`DiagCtxHandle::drain()`].
    fn drain(&mut self) {
        let encountered_errors = match self.handler.drain() {
            Ok(_) => return,
            Err(error_snippet::DrainError::Fmt(e)) => panic!("{:?}", e),
            Err(error_snippet::DrainError::CompoundError(cnt)) => cnt,
        };

        let message = format!("aborting due to {} previous errors", encountered_errors);
        let abort_diag = Box::new(SimpleDiagnostic::new(message));

        let _ = self.handler.report_and_drain(abort_diag);

        if self.exit_on_error {
            // Raise a fatal error, without printing a backtrace.
            FatalError::raise();
        } else {
            // Mark the context as tainted.
            self.tainted.store(true, Ordering::Release);
        }
    }
}

/// A context to deal with diagnostics, which is meant to
/// be used throughout the entire lifespan of the compiler / driver
/// process.
///
/// Certain diagnostics may cause a single stage within the compiler
/// to halt or exit early, where-as others might be more benign.
pub struct DiagCtx {
    /// The inner handler for diagnostics, which holds all the
    /// reporting diagnostics.
    inner: Arc<Mutex<DiagCtxInner>>,
}

impl DiagCtx {
    /// Creates a new [`DiagCtx`] instance using the given output format.
    pub fn new(fmt: DiagOutputFormat) -> Self {
        let inner = DiagCtxInner::new(fmt);

        DiagCtx {
            inner: Arc::new(Mutex::new(inner)),
        }
    }

    /// Retrives the instance of the parent [`DiagCtxInner`], which
    /// is contained within the context.
    fn handler(&self) -> MutexGuard<'_, DiagCtxInner> {
        self.inner.lock().unwrap()
    }

    /// Enables the context to halt the application, if an error is encountered
    /// during a drain.
    pub fn exit_on_error(&mut self) {
        self.handler().exit_on_error = true;
    }

    /// Create a handle for the diagnostic context, which can be
    /// used to emit diagnositcs to the inner context.
    pub fn handle(&mut self) -> DiagCtxHandle {
        DiagCtxHandle {
            inner: Arc::clone(&self.inner),
        }
    }

    /// Creates a new handle, which is only valid within the given closure,
    /// which is executed immedietly. Upon finishing the closure, the handle is dropped
    /// and all diagnostics reporting within it are immedietly drained to the inner handler.
    pub fn with<TReturn>(&mut self, mut f: impl FnMut(DiagCtxHandle) -> TReturn) -> TReturn {
        let handle = self.handle();

        f(handle)
    }

    /// Determines whether the diagnostic context is tainted.
    ///
    /// The context can become tainted if one-or-more errors are reporting
    /// during a drain to the parent handler. This does not include warnings or
    /// sub-diagnostics.
    ///
    /// # Examples
    ///
    /// Using [`DiagCtx::with()`]:
    /// ```
    /// use error_snippet::SimpleDiagnostic;
    /// use lume_errors::{DiagCtx, DiagOutputFormat};
    ///
    /// let mut dcx = DiagCtx::new(DiagOutputFormat::Stubbed);
    ///
    /// dcx.with(|mut handle| {
    ///     let error = SimpleDiagnostic::new("An error occurred");
    ///     handle.emit(error.into());
    /// });
    ///
    /// assert_eq!(dcx.tainted(), true);
    /// ```
    ///
    /// Errors emitted to the [`DiagCtxHandle`] handle are only drained when
    /// calling the [`DiagCtxHandle::drain()`] method manually, or when the handle
    /// is dropped. So, the following example will *not* taint the context:
    /// ```
    /// use error_snippet::SimpleDiagnostic;
    /// use lume_errors::{DiagCtx, DiagOutputFormat};
    ///
    /// let mut dcx = DiagCtx::new(DiagOutputFormat::Stubbed);
    /// let mut handle = dcx.handle();
    ///
    /// let error = SimpleDiagnostic::new("An error occurred");
    /// handle.emit(error.into());
    ///
    /// // `handle` is not dropped, so no errors are drained!
    /// assert_eq!(dcx.tainted(), false);
    ///
    /// // we can manually drain using [`DiagCtxHandle::drain()`] or
    /// // simply drop the handle.
    /// drop(handle);
    /// assert_eq!(dcx.tainted(), true);
    /// ```
    pub fn tainted(&self) -> bool {
        self.handler().tainted.load(Ordering::Acquire)
    }
}

/// A handle to a parent [`DiagCtx`], which can be used in
/// distinct sequential "stages", where each stage can only progress
/// forward if no halting diagnostics were reporting in any of the previous
/// stages.
///
/// The handle acts as a mutable reference to it's parent [`DiagCtx`] instance,
/// but will drain all errors to the output, once it's been dropped or manually drained.
pub struct DiagCtxHandle {
    /// Contains the parent [`DiagCtxInner`] handler.
    inner: Arc<Mutex<DiagCtxInner>>,
}

impl DiagCtxHandle {
    /// Creates a new [`DiagCtxHandle`] with a stubbed renderer, functioning
    /// similar to a shim. Mostly used for testing.
    pub fn shim() -> Self {
        DiagCtx::new(DiagOutputFormat::Stubbed).handle()
    }

    /// Retrives the instance of the parent [`DiagCtxInner`], which
    /// is contained within the handle.
    fn handler(&self) -> MutexGuard<'_, DiagCtxInner> {
        self.inner.lock().unwrap()
    }

    /// Emits the given error to the parent context.
    ///
    /// Depending on the severity of the error, it might cause the
    /// execution of the program to halt (e.g. fatal errors, bugs, etc.), but
    /// most would simply be reporting until drained from the context.
    #[track_caller]
    pub fn emit(&mut self, diagnostic: error_snippet::Error) {
        self.handler().handler.report(diagnostic);
    }

    /// Drains the currently reported errors in the context to the output buffer.
    ///
    /// If any reported diagnostics have a severity at or above [`error_snippet::Severity::Error`],
    /// they will be counted towards a [`error_snippet::DrainError::CompoundError`], which will be
    /// raised when draining has finished.
    pub fn drain(&mut self) {
        self.handler().drain();
    }

    /// Determines whether the diagnostic context is tainted.
    ///
    /// For more information, read the documentation on [`DiagCtx::tainted()`].
    pub fn tainted(&self) -> bool {
        self.handler().tainted.load(Ordering::Acquire)
    }
}

unsafe impl Send for DiagCtxHandle {}
unsafe impl Sync for DiagCtxHandle {}

impl Drop for DiagCtxHandle {
    fn drop(&mut self) {
        self.drain();
    }
}

/// Defines a [`Renderer`] for handlers, which don't report any errors.
/// Instead, it sends the reported diagnostics to the void, where they won't disturb anyone.
struct StubRenderer {}

impl Renderer for StubRenderer {
    fn render_fmt(
        &mut self,
        _f: &mut error_snippet::Formatter,
        _diagnostic: &dyn error_snippet::Diagnostic,
    ) -> std::fmt::Result {
        Ok(())
    }
}
