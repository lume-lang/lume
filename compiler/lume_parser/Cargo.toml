[package]
name = "lume_parser"
version.workspace = true
edition = "2024"
rust-version.workspace = true

[dependencies]
lume_ast = { path = "../lume_ast" }
lume_diag = { path = "../lume_diag" }
lume_diag_macros = { path = "../lume_diag_macros" }
lume_span = { path = "../lume_span" }
lume_state = { path = "../lume_state" }

serde = { version = "1.0.219", features = ["serde_derive"] }

[dev-dependencies]
insta = { version = "1.42.2", features = ["yaml"] }
