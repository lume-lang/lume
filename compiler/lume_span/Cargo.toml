[package]
name = "lume_span"
version.workspace = true
edition = "2024"
rust-version.workspace = true

[dependencies]
fxhash = "^0.2"
indexmap = { version = "^2.9", features = ["serde"] }
serde = { version = "1.0.219", features = ["serde_derive"] }
