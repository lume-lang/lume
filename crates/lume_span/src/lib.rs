//! Source files and spans within them, which are used heavily within the Lume compiler.
//!
//! This module is used by most other packages within the Lume compiler, since spans
//! are required to print useful diagnostics to the user - at least if source code is needed.

pub mod id;
pub mod intern;
pub mod source;

use std::{hash::Hash, sync::Arc};

pub use id::*;
pub use intern::*;
pub use source::{FileName, SourceFile, SourceFileId, SourceMap};

/// Hashes some ID using the `FxHasher` algorithm, which was extracted
/// from the Rustc compiler.
///
/// The reason for using this instead of the [`std::hash::DefaultHasher`] is that
/// we require some deterministic hashing algorithm for consistent results, so
/// we can use it for incremental builds and caching. The default hasher does not
/// have any specific algorithm defined, so it cannot be relied upon to create
/// the same hash given the same input.
pub fn hash_id<T: Hash + ?Sized>(id: &T) -> usize {
    fxhash::hash(id)
}

/// Represents some marked location within a source file.
///
/// This structure is used heavily throughout the compiler, to define
/// locations of statements and expressions, which are used to create more useful
/// diagnostics and error messages.
pub type Location = Interned<source::Location>;

impl Location {
    pub fn empty() -> Self {
        source::Location::empty().intern()
    }
}

impl From<Location> for Arc<dyn error_snippet::Source> {
    fn from(value: Location) -> Self {
        value.file.clone()
    }
}

impl From<Location> for error_snippet::SpanRange {
    fn from(value: Location) -> Self {
        error_snippet::SpanRange(value.index.start..value.index.end)
    }
}

impl Default for Location {
    fn default() -> Self {
        Self::empty()
    }
}
