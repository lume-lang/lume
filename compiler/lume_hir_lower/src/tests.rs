use lume_errors::{DiagCtx, DiagOutputFormat};
use lume_parser::Parser;
use lume_span::PackageId;

use super::*;

#[track_caller]
fn lower(input: &str) -> Result<Map> {
    let dcx = DiagCtx::new(DiagOutputFormat::Stubbed);
    let source = Arc::new(SourceFile::internal(input));

    let expressions = dcx
        .with_res(|handle| Parser::new(source.clone(), handle))?
        .parse()
        .unwrap();

    let module_id = PackageId::empty();
    let mut map = Map::empty(module_id);

    dcx.with(|handle| LowerModule::lower(&mut map, source, handle, expressions))?;

    Ok(map)
}

#[track_caller]
fn lower_expr(input: &str) -> Result<Vec<lume_hir::Statement>> {
    let dcx = DiagCtx::new(DiagOutputFormat::Stubbed);
    let source = Arc::new(SourceFile::internal(input));

    let mut parser = Parser::new_with_str(input);
    parser.prepare()?;

    let statements = parser.parse_statements()?;

    let module_id = PackageId::empty();
    let mut map = Map::empty(module_id);
    let mut lower = dcx.with_res(|handle| LowerModule::new(&mut map, source, handle))?;

    Ok(lower.statements(statements))
}

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

macro_rules! assert_snap_eq {
    (
        $input: expr,
        $($expr:expr),+
    ) => {
        set_snapshot_suffix!( $($expr),+ );

        insta::with_settings!({
            description => $input,
            omit_expression => true
        }, {
            insta::assert_debug_snapshot!(lower($input).unwrap());
        });
    };
}

macro_rules! assert_err_snap_eq {
    (
        $input: expr,
        $($expr:expr),+
    ) => {
        set_snapshot_suffix!( $($expr),+ );

        insta::with_settings!({
            description => $input,
            omit_expression => true
        }, {
            insta::assert_debug_snapshot!(lower($input).unwrap_err());
        });
    };
}

macro_rules! assert_expr_snap_eq {
    (
        $input: expr,
        $($expr:expr),+
    ) => {
        set_snapshot_suffix!( $($expr),+ );

        insta::with_settings!({
            description => $input,
            omit_expression => true
        }, {
            insta::assert_debug_snapshot!(lower_expr($input).unwrap());
        });
    };
}

#[test]
fn test_implicit_imports() {
    for ty in &[
        "Boolean",
        "String",
        "Int8",
        "UInt8",
        "Int16",
        "UInt16",
        "Int32",
        "UInt32",
        "Int64",
        "UInt64",
        "Float",
        "Double",
        "Array<Int32>",
        "Pointer",
        "Range<Int32>",
        "RangeInclusive<Int32>",
    ] {
        assert_snap_eq!(&format!("fn foo() -> {ty} {{ }}"), "{}", ty);
    }
}

#[test]
fn test_function_definition_snapshots() {
    assert_snap_eq!("fn main() -> void {}", "empty");
    assert_snap_eq!("fn main() -> void { let a = 0; }", "statement");
    assert_snap_eq!("fn main() -> void { let a = 0; let b = 1; }", "statements");
    assert_snap_eq!("fn main() {}", "no_return_type");
    assert_snap_eq!("fn main(argc: u8) -> void { }", "parameter");
    assert_snap_eq!("fn main(argc: u8, arcv: [String]) -> void { }", "parameters");
    assert_snap_eq!("fn external main() -> void", "external");
    assert_snap_eq!("pub fn main() -> void {}", "pub_modifier");
    assert_snap_eq!("fn loop() -> void {}", "reserved_keyword");
    assert_snap_eq!("fn main() -> std::Int32 {}", "namespaced_type");
    assert_snap_eq!("fn empty?() {}", "boolean_function");
    assert_snap_eq!("fn foo(...args: Int32) {}", "varargs");
    assert_err_snap_eq!("fn foo(...args: Int32, a: Int32) {}", "varargs_nonlast");
}

#[test]
fn test_cast_snapshots() {
    assert_expr_snap_eq!("0 as u64;", "literal_single");
    assert_expr_snap_eq!("let _ = 0 as u64;", "precedence_assign");
    assert_expr_snap_eq!("0 as u64 == 0;", "precedence_equal");
    assert_expr_snap_eq!("call() as u64;", "invocation");
}

#[test]
fn test_binary_snapshots() {
    assert_expr_snap_eq!("0 & 0;", "and");
    assert_expr_snap_eq!("0 | 0;", "or");
    assert_expr_snap_eq!("0 ^ 0;", "xor");
}

#[test]
fn test_boolean_snapshots() {
    assert_expr_snap_eq!("0 && 0;", "and");
    assert_expr_snap_eq!("0 || 0;", "or");
}

#[test]
fn test_literal_snapshots() {
    assert_expr_snap_eq!("\"\";", "string_empty");
    assert_expr_snap_eq!("\"string\";", "string_content");
    assert_expr_snap_eq!("true;", "bool_true");
    assert_expr_snap_eq!("false;", "bool_false");
    assert_expr_snap_eq!("let ident = 0;", "ident");
    assert_expr_snap_eq!("let IDENT = 0;", "ident_case");
    assert_expr_snap_eq!("let __IDENT__ = 0;", "ident_underscore");
    assert_expr_snap_eq!("0;", "int");
    assert_expr_snap_eq!("55;", "int_positive");
    assert_expr_snap_eq!("-55;", "int_negative");
    assert_expr_snap_eq!("0x55;", "int_hex_positive");
    assert_expr_snap_eq!("-0x55;", "int_hex_negative");
    assert_expr_snap_eq!("0b01010101;", "int_bin_positive");
    assert_expr_snap_eq!("-0b01010101;", "int_bin_negative");
    assert_expr_snap_eq!("0o125;", "int_oct_positive");
    assert_expr_snap_eq!("-0o125;", "int_oct_negative");
}

#[test]
fn test_conditional_snapshots() {
    assert_expr_snap_eq!("if true { }", "if_empty");
    assert_expr_snap_eq!("if true { let a = 1; }", "if_statement");
    assert_expr_snap_eq!("if true { } else if false { }", "if_else_if_empty");
    assert_expr_snap_eq!("if true { } else { }", "if_else_empty");
    assert_expr_snap_eq!("if true { } else if false { } else { }", "if_else_if_else_empty");
    assert_expr_snap_eq!("unless true { }", "unless_empty");
    assert_expr_snap_eq!("unless true { } else { }", "unless_else_empty");
    assert_expr_snap_eq!("let a = 0; if a == 1 { }", "equality_empty");
    assert_expr_snap_eq!("let a = 0; if a != 1 { }", "inequality_empty");
    assert_expr_snap_eq!("if true { let a = 0; }", "if_statement");
    assert_expr_snap_eq!("if true { let a = 0; let b = 0; }", "if_statements");
    assert_expr_snap_eq!(
        "if true { let a = 0; } else if false { let a = 0; }",
        "else_if_statements"
    );
    assert_expr_snap_eq!("unless true { let a = 0; }", "unless_statement");
}

#[test]
fn test_loop_snapshots() {
    assert_expr_snap_eq!("loop { }", "inf_loop_empty");
    assert_expr_snap_eq!("loop { let a = 0; }", "inf_loop_statement");
    assert_expr_snap_eq!("loop { break; }", "inf_loop_break");
    assert_expr_snap_eq!("loop { continue; }", "inf_loop_continue");

    assert_expr_snap_eq!("while true { }", "pred_loop_empty");
    assert_expr_snap_eq!("while true { let a = 0; }", "pred_loop_statement");
    assert_expr_snap_eq!("while true { break; }", "pred_loop_break");
    assert_expr_snap_eq!("while true { continue; }", "pred_loop_continue");

    assert_expr_snap_eq!("for pattern in [1, 2, 3] { }", "iter_loop_empty");
    assert_expr_snap_eq!("for pattern in [1, 2, 3] { let a = 0; }", "iter_loop_statement");
    assert_expr_snap_eq!("for pattern in [1, 2, 3] { break; }", "iter_loop_break");
    assert_expr_snap_eq!("for pattern in [1, 2, 3] { continue; }", "iter_loop_continue");

    assert_expr_snap_eq!(
        "
        let collection = [1, 2, 3];
        for pattern in collection { }",
        "iter_loop_empty_var"
    );

    assert_expr_snap_eq!(
        "
        let collection = [1, 2, 3];
        for pattern in collection { let a = 0; }",
        "iter_loop_statement_var"
    );

    assert_expr_snap_eq!(
        "
        let collection = [1, 2, 3];
        for pattern in collection { break; }",
        "iter_loop_break_var"
    );

    assert_expr_snap_eq!(
        "
        let collection = [1, 2, 3];
        for pattern in collection { continue; }",
        "iter_loop_continue_var"
    );
}

#[test]
fn test_range_snapshots() {
    assert_expr_snap_eq!("let _ = (0..1);", "literal_exclusive");
    assert_expr_snap_eq!("let _ = (0..=1);", "literal_inclusive");
    assert_expr_snap_eq!(
        "
        let a = 0;
        let b = 1;
        let _ = (a..b);",
        "expr_exclusive"
    );

    assert_expr_snap_eq!(
        "
        let a = 0;
        let b = 1;
        let _ = (a..=b);",
        "expr_inclusive"
    );

    assert_expr_snap_eq!(
        "
        let a = 0;
        let b = 1;
        let _ = ((a + b)..(a + b + 1));",
        "expr_nested_exclusive"
    );

    assert_expr_snap_eq!(
        "
        let a = 0;
        let b = 1;
        let _ = ((a + b)..=(a + b + 1));",
        "expr_nested_inclusive"
    );
}

#[test]
fn test_call_snapshots() {
    assert_expr_snap_eq!("let _ = call();", "function_empty");
    assert_expr_snap_eq!("let _ = call(0);", "function_param_1");
    assert_expr_snap_eq!("let _ = call(0, 1);", "function_param_2");
    assert_expr_snap_eq!("let _ = call::<T>(0, 1);", "function_generic");

    assert_expr_snap_eq!("let _ = 1.call();", "method_empty");
    assert_expr_snap_eq!("let _ = 1.call(0);", "method_param_1");
    assert_expr_snap_eq!("let _ = 1.call(0, 1);", "method_param_2");
    assert_expr_snap_eq!("let _ = 1.call::<T>(0, 1);", "method_generic");
    assert_expr_snap_eq!("let _ = Int32::<T>::call(0, 1);", "generic_static_method");
}

#[test]
fn test_return_snapshots() {
    assert_expr_snap_eq!("return;", "empty");
    assert_expr_snap_eq!("return 1;", "scalar");
    assert_expr_snap_eq!(
        "
        let a = 1;
        let b = 1;
        let c = 1;

        return a.b(c);",
        "call"
    );
}

#[test]
fn test_generic_function_snapshots() {
    assert_snap_eq!("fn test() -> void {}", "no_generics");
    assert_snap_eq!("fn test<>() -> void {}", "empty_generics");
    assert_snap_eq!("fn test<T>() -> void {}", "single_generic");
    assert_snap_eq!("fn test<T1, T2>() -> void {}", "multiple_generics");
    assert_snap_eq!("fn test<T: Numeric>() -> void {}", "constrained_generic");
    assert_snap_eq!("fn test<T1: Numeric, T2: Numeric>() -> void {}", "constrained_generics");
    assert_snap_eq!(
        "fn test<T: Numeric + Floating>() -> void {}",
        "multiple_constrained_types"
    );
    assert_snap_eq!(
        "fn test<T1: Numeric + Floating, T2: Numeric + Floating>() -> void {}",
        "multiple_types_with_multiple_constraints"
    );
}

#[test]
fn test_struct_snapshots() {
    assert_snap_eq!("struct Int32 {}", "empty_decl");
    assert_snap_eq!("impl Int32 {}", "empty_impl");

    assert_snap_eq!(
        "
            impl Foo {
                fn bar() -> Int32 {
                    return 0;
                }
            }",
        "method"
    );

    assert_snap_eq!(
        "
            impl Foo {
                fn bar() { }
            }",
        "method_no_ret"
    );

    assert_snap_eq!(
        "
            impl Foo {
                pub fn bar() -> Int32 {
                    return 0;
                }
            }",
        "pub_method"
    );

    assert_snap_eq!(
        "
            impl Foo {
                fn external bar() -> Int32
            }",
        "ext_method"
    );

    assert_snap_eq!(
        "
            impl Foo {
                pub fn ==() -> bool {
                    return true;
                }
            }",
        "operator_method"
    );

    assert_snap_eq!(
        "
            impl Foo {
                fn bar<T>() -> Int32 { }
            }",
        "generic_method"
    );

    assert_snap_eq!(
        "
            struct Foo {
                x: Int32 = 0;
            }",
        "property"
    );

    assert_snap_eq!(
        "
            struct Foo {
                x: Int32;
            }",
        "property_no_default"
    );

    assert_snap_eq!(
        "
            struct Foo {
                pub x: Int32 = 1;
            }",
        "pub_property"
    );
}

#[test]
fn test_generic_struct_snapshots() {
    assert_snap_eq!("struct Test {}", "no_generics");
    assert_snap_eq!("struct Test<> {}", "empty_generics");
    assert_snap_eq!("struct Test<T> {}", "single_generic");
    assert_snap_eq!("struct Test<T1, T2> {}", "multiple_generics");
    assert_snap_eq!("struct Test<T: Numeric> {}", "constrained_generic");
    assert_snap_eq!("struct Test<T1: Numeric, T2: Numeric> {}", "constrained_generics");
}

#[test]
fn test_generic_method_snapshots() {
    assert_snap_eq!("impl Test { fn test() -> void {} }", "no_generics");
    assert_snap_eq!("impl Test { fn test<>() -> void {} }", "empty_generics");
    assert_snap_eq!("impl Test { fn test<T>() -> void {} }", "single_generic");
    assert_snap_eq!("impl Test { fn test<T1, T2>() -> void {} }", "multiple_generics");
    assert_snap_eq!("impl Test { fn test<T: Numeric>() -> void {} }", "constrained_generic");
    assert_snap_eq!(
        "impl Test { fn test<T1: Numeric, T2: Numeric>() -> void {} }",
        "constrained_generics"
    );
}

#[test]
fn test_enum_snapshots() {
    assert_snap_eq!("enum Foo {}", "empty");
    assert_snap_eq!("enum Foo { Bar }", "single_variant");
    assert_snap_eq!("enum Foo { Bar, Baz }", "multiple_variants");

    assert_snap_eq!(
        "
            enum Foo {
                Bar()
            }",
        "variant_param_empty"
    );

    assert_snap_eq!(
        "
            enum Foo {
                Bar(int)
            }",
        "variant_param_single"
    );

    assert_snap_eq!(
        "
            enum Foo {
                Bar(int, int)
            }",
        "variant_param_multiple"
    );

    assert_snap_eq!(
        "
            enum Foo {
                Bar(int, int),
                Baz(int, int)
            }",
        "multiple_variants_multiple_params"
    );
}

#[test]
fn test_type_alias_snapshots() {
    assert_snap_eq!("type A = B", "scalar");
    assert_snap_eq!("type A = [B]", "array");
    assert_snap_eq!("type A = B<C>", "generic");
}

#[test]
fn test_trait_snapshots() {
    assert_snap_eq!("trait Add { }", "empty");
    assert_snap_eq!("trait Add { pub fn add(other: int) -> int; }", "method");
    assert_snap_eq!("trait Add { pub fn add(other: int) -> int { } }", "method_impl");
    assert_snap_eq!("trait Add<T> { }", "generic");
    assert_snap_eq!("trait Add<T1, T2> { }", "generics");
    assert_snap_eq!("trait Add { fn add(other: int) -> int; }", "private_method");
    assert_snap_eq!("trait Add<T: Numeric> {}", "constrained_generic");
    assert_snap_eq!("trait Add<T1: Numeric, T2: Numeric> {}", "constrained_generics");
    assert_snap_eq!("trait Add { pub fn add(other: int) { } }", "method_no_ret");
}

#[test]
fn test_use_trait_snapshots() {
    assert_snap_eq!("use Add in Int32 {}", "empty");

    assert_snap_eq!(
        "
            use Add in Int32 {
                fn add(other: Int32) -> Int32 {}
            }",
        "priv_method"
    );

    assert_snap_eq!(
        "
            use Add in Int32 {
                pub fn add(other: Int32) -> Int32 {}
            }",
        "pub_method"
    );

    assert_snap_eq!(
        "
            use Add in Int32 {
                fn add(other: Int32) {}
            }",
        "method_no_ret"
    );

    assert_snap_eq!(
        "
            use Cast in Int32 {
                pub fn to_string() -> String {}

                pub fn to_int() -> Int32 {}
            }",
        "methods"
    );

    assert_snap_eq!(
        "
            use Add<Int32> in Int32 {
                pub fn add(other: Int32) -> Int32 {}
            }",
        "generic"
    );

    assert_snap_eq!(
        "
            use Add<Int32, Int64> in Int32 {
                pub fn add(other: Int32) -> Int64 {}
            }",
        "generics"
    );

    assert_snap_eq!(
        "
            use Enumerable<T> in Vector<T> {
                pub fn next() -> T {}
            }",
        "generic_type"
    );

    assert_snap_eq!(
        "
            use<T> Enumerable<T> in Vector<T> {
                pub fn next() -> T { }
            }",
        "generic_trait"
    );
}

#[test]
fn test_doc_comments_snapshots() {
    assert_snap_eq!(
        "/// This is a doc comment
        fn foo() -> void { }",
        "function"
    );

    assert_snap_eq!(
        "/// This is a doc comment
        struct Foo { }",
        "struct"
    );

    assert_snap_eq!(
        "struct Foo {
            /// This is a doc comment
            pub bar: Int32 = 0;
        }",
        "property"
    );

    assert_snap_eq!(
        "impl Foo {
            /// This is a doc comment
            pub fn bar() -> void { }
        }",
        "method"
    );

    assert_snap_eq!(
        "/// This is a doc comment
        trait Foo { }",
        "trait"
    );

    assert_snap_eq!(
        "trait Foo {
            /// This is a doc comment
            pub fn bar() -> void { }
        }",
        "trait_method"
    );

    assert_snap_eq!(
        "/// This is a doc comment
        enum Foo {
            Bar
        }",
        "enum"
    );

    assert_snap_eq!(
        "enum Foo {
            /// This is a doc comment
            Bar
        }",
        "enum_case"
    );

    assert_snap_eq!(
        "/// This is a doc comment
        type Foo = Bar",
        "type_alias"
    );
}

#[test]
fn test_type_aliasing() {
    assert_snap_eq!("type A = B", "scalar_to_scalar");
    assert_snap_eq!("type A = [B]", "scalar_to_array");
    assert_snap_eq!("type A = B<C>", "scalar_to_generic");
}

#[test]
fn test_using_self_as_parameter() {
    assert_snap_eq!(
        "impl Foo {
        pub fn bar(self, a: Int32) -> void { }
    }",
        "valid"
    );

    assert_err_snap_eq!(
        "impl Foo {
        pub fn bar(a: Int32, self) -> void { }
    }",
        "invalid"
    );
}

#[test]
fn test_using_self_in_function() {
    assert_snap_eq!("fn foo() -> void { }", "valid");

    assert_err_snap_eq!("fn foo(self) -> void { }", "invalid");
}
