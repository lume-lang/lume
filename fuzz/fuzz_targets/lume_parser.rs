#![no_main]

use std::sync::Arc;

use libfuzzer_sys::fuzz_target;
use lume_errors::DiagCtxHandle;
use lume_span::SourceFile;

fuzz_target!(|data: &[u8]| {
    if let Ok(s) = std::str::from_utf8(data) {
        let source = Arc::new(SourceFile::internal(s));

        let mut lexer = lume_lexer::Lexer::new(source.clone());
        let tokens = lexer.lex().unwrap();

        let mut parser = lume_parser::Parser::new(source, tokens, DiagCtxHandle::shim()).unwrap();
        parser.disable_recovery();

        let _ = parser.parse();
    }
});
