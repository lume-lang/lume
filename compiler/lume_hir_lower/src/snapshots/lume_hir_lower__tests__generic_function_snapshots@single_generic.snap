---
source: compiler/lume_hir_lower/src/tests.rs
description: "fn test<T>() -> void {}"
---
Map {
    package: PackageId(
        0,
    ),
    items: {
        ItemId(
            14825577416135075223,
        ): Function(
            FunctionDefinition {
                id: ItemId(
                    14825577416135075223,
                ),
                func_id: None,
                visibility: Private,
                name: SymbolName {
                    namespace: None,
                    name: Identifier {
                        name: "test",
                        location: <internal>:3:7,
                    },
                    location: <internal>:3:7,
                },
                parameters: [],
                type_parameters: [
                    TypeParameter {
                        name: Identifier {
                            name: "T",
                            location: <internal>:8:9,
                        },
                        type_id: None,
                        type_param_id: None,
                        constraints: [],
                        location: <internal>:8:9,
                    },
                ],
                return_type: Some(
                    Type {
                        id: ItemId(
                            9344623788042179276,
                        ),
                        name: SymbolName {
                            namespace: Some(
                                PathRoot {
                                    segments: [],
                                },
                            ),
                            name: Identifier {
                                name: "void",
                                location: <internal>:16:20,
                            },
                            location: <internal>:16:20,
                        },
                        type_params: [],
                        location: <internal>:16:20,
                    },
                ),
                block: Some(
                    Block {
                        statements: [],
                        location: <internal>:21:23,
                    },
                ),
                location: <internal>:0:23,
            },
        ),
    },
    statements: {},
    expressions: {},
}
